packages/typography/src/Link/Link.tsx

```diff
@@ -0,0 +1,100 @@
+import React, { useEffect, useMemo, useState } from 'react';
+import ArrowRightIcon from '@leafygreen-ui/icon/dist/ArrowRight';
+import OpenNewTabIcon from '@leafygreen-ui/icon/dist/OpenNewTab';
+import Box, { ExtendableBox } from '@leafygreen-ui/box';
+import { cx } from '@leafygreen-ui/emotion';
+import { useDarkMode } from '@leafygreen-ui/leafygreen-provider';
+import { bodyTypeScaleStyles } from '../styles';
+import { useUpdatedBaseFontSize } from '../utils/useUpdatedBaseFontSize';
+import {
+  openInNewTabStyles,
+  arrowRightIconHover,
+  arrowRightIconPersist,
+  anchorClassName,
+  linkStyles,
+  linkModeStyles,
+  underlineModeStyles,
+  underlineStyles,
+} from './Link.styles';
+import { LinkProps, ArrowAppearance } from './Link.types';
+
+const Link: ExtendableBox<LinkProps, 'a'> = ({
+  href,
+  children,
+  className,
+  target: targetProp,
+  arrowAppearance = ArrowAppearance.None,
+  hideExternalIcon = false,
+  darkMode: darkModeProp,
+  ...rest
+}: LinkProps) => {
+  const [currentHostname, setCurrentHostname] = useState('');
+  useEffect(() => {
+    setCurrentHostname(window.location.hostname);
+  }, []);
+
+  const { theme } = useDarkMode(darkModeProp);
+
+  const hrefHostname = useMemo(() => {
+    if (!href) return;
+    const httpRegex = /^http(s)?:\/\//;
+    return httpRegex.test(href) ? new URL(href).hostname : currentHostname;
+  }, [href, currentHostname]);
+
+  const baseFontSize = useUpdatedBaseFontSize();
+
+  let target, icon, rel;
+
+  if (targetProp) {
+    target = targetProp;
+  } else {
+    if (hrefHostname === currentHostname) {
+      target = '_self';
+    } else {
+      target = '_blank';
+      rel = 'noopener noreferrer';
+    }
+  }
+
+  if (target === '_blank' && !hideExternalIcon) {
+    icon = (
+      <OpenNewTabIcon role="presentation" className={openInNewTabStyles} />
+    );
+  } else if (arrowAppearance !== ArrowAppearance.None) {
+    icon = (
+      <ArrowRightIcon
+        role="presentation"
+        size={12}
+        className={cx({
+          [arrowRightIconHover]: arrowAppearance === ArrowAppearance.Hover,
+          [arrowRightIconPersist]: arrowAppearance === ArrowAppearance.Persist,
+        })}
+      />
+    );
+  }
+
+  const elementProps = href
+    ? ({ as: 'a', href, target, rel } as const)
+    : ({ as: 'span' } as const);
```

Comment:
There's a lot going on in this component. We could reorganize a lot of this (i.e. a `LinkIcon` subcomponent)

Reply:
Goal of this PR was purely structural -- don't want code changes muddying the ability to merge.